
---
title: 'Demo of DIRECT-NET using parallel scRNA-seq and scATAC-seq data of human A549 cell line'
author: "Lihua Zhang"
date: "`r format(Sys.time(), '%d/%m/%y')`"
output:
  html_document: default
mainfont: Arial
vignette: |
  %\VignetteIndexEntry{Inferring CREs and reconstructing regulatory networks from single cell multi-omics data} %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  root.dir = './'
)
```

We showcase DIRECT-NETâ€™s capability of inferring CREs by applying it to parallel scRNA-seq and scATAC-seq data of three time points of DEX-treatmented human A549 cell line. 

To make it easy to run DIRECT-NET in most common parallel scRNA-seq and scATAC-seq data or scATAC-seq data analysis pipelines, the implementent of DIRECT-NET is seamlessly compatible with the workflow of Seurat/Signac package. In sum, DIRECT-NET pipeline consists of the following four major parts:

* Set up a Seurat object

* Infer links using DIRECT-NET by taking an input a Seurat object

* Visualize the links

* Construct regulatory networks

DIRECT-NET's workflow closely follows the Seurat vignette: [https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html](https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html)

## Load the required libraries
```{r message=FALSE,warning=FALSE}
library(DIRECTNET)
library(Seurat)
library(Signac)
library(EnsDb.Hsapiens.v86)
library(patchwork)
library(dplyr)
library(ggplot2)
options(stringsAsFactors = FALSE)
```

## Load data
The scRNA and scATAC-seq datasets we demonstrated here, including two count data matrices for three time points, can be downloaded via this shared Google Drive [link](https://drive.google.com/drive/folders/13KyLEjJ6iOm6_HJYoP2zK_nv-juHLmdT?usp=sharing).

```{r}
rna.counts <- read.table(file = "/Users/lihuazhang/Documents/DIRECT-NET/A549/Data/A549_RNA_count.txt")
atac.counts <- read.table(file = "/Users/lihuazhang/Documents/DIRECT-NET/A549/Data/A549_ATAC_count.txt")
label <- read.table("/Users/lihuazhang/Documents/DIRECT-NET/A549/Data/A549_time.txt",sep = "\t")
names(label) <- c("Cells","time")
meta.data <- as.character(label$time)
meta.data <- as.data.frame(meta.data)
rownames(meta.data) <- as.character(label$Cells)
colnames(meta.data) <- "celltype"

markers <- read.table(file = "/Users/lihuazhang/Documents/DIRECT-NET/A549/Data/A549_DE_genes.txt",header = TRUE,sep = "\t") # identified by Wilcox rank test
markers$gene <- as.character(markers$gene)
markers$group <- as.character(markers$group)

genome.info <- read.table(file = "/Users/lihuazhang/Documents/DIRECT-NET/A549/Data/hg19.promoter.regions.txt")
names(genome.info) <- c("Chrom","Starts","Ends","genes")
genes <- lapply(genome.info$genes, function(x) strsplit(x,"[|]")[[1]][1])
genes <- lapply(genes, function(x) strsplit(x,"[.]")[[1]][1])
genes <- unlist(genes)
genome.info$genes <- genes
unik <- !duplicated(genes)# filter out different transcript
genome.info <- genome.info[unik,]

peak <- rownames(atac.counts)
peak1 <- strsplit(peak,"-")
peaks <- matrix(0, nrow = length(peak),ncol = 1)
for (i in 1:length(peaks)) {
  peaks[i] <- paste0("chr",peak1[[i]][1],":",peak1[[i]][2],"-",peak1[[i]][3])
}
rownames(atac.counts) <- as.vector(peaks)

```

# Part I: Set up a Seurat object
```{r, message= FALSE, warning=FALSE}
combined <- CreateSeuratObject(counts = rna.counts, meta.data = meta.data) 

# Now add in the ATAC-seq data
# we'll only use peaks in standard chromosomes
grange.counts <- StringToGRanges(rownames(atac.counts), sep = c(":", "-"))
grange.use <- seqnames(grange.counts) %in% standardChromosomes(grange.counts)
atac.counts <- atac.counts[as.vector(grange.use), ]
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotations) <- 'UCSC'
genome.name <- "hg19"
genome(annotations) <- genome.name

chrom_assay <- CreateChromatinAssay(
  counts = atac.counts,
  sep = c(":", "-"),
  genome = genome.name,
  min.cells = 0,
  annotation = annotations
)
combined[["ATAC"]] <- chrom_assay

```

###  RNA analysis
```{r, warning=FALSE}
DefaultAssay(combined) <- "RNA"
combined <- NormalizeData(object = combined, normalization.method = "LogNormalize", scale.factor = 10000)
# Detection of variable genes across the single cells
combined <- FindVariableFeatures(object = combined)
features.rna <- markers$gene
combined@assays$RNA@var.features <- features.rna
combined <- ScaleData(object = combined)
combined <- RunPCA(combined, pc.genes = combined@var.genes, pcs.compute = 40, do.print = FALSE)
combined <- RunUMAP(combined, dims = 1:40)

```

### ATAC analysis
```{r}
# We exclude the first dimension as this is typically correlated with sequencing depth
DefaultAssay(combined) <- "ATAC"
combined <- RunTFIDF(combined)
combined <- FindTopFeatures(combined, min.cutoff = 'q0')
combined <- RunSVD(combined)
combined <- RunUMAP(combined, reduction = 'lsi', dims = 2:50, reduction.name = "umap.atac", reduction.key = "atacUMAP_")
combined <- FindMultiModalNeighbors(combined, reduction.list = list("pca", "lsi"), dims.list = list(1:50, 2:50))
combined <- RunUMAP(combined, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
Idents(combined) <- combined$celltype

```

# Part II: Infer CREs using DIRECT-NET
```{r,message=FALSE,warning=FALSE}
# Infer links for focused markers
combined <- Run_DIRECT_NET(combined, k_neigh = 50, atacbinary = TRUE, max_overlap=0.5, size_factor_normalize = TRUE, genome.info = genome.info, focus_markers = c("PYGB","NR3C1"))
direct.net_result <- Misc(combined, slot = 'direct.net')
direct.net_result <- as.data.frame(do.call(cbind,direct.net_result))

```

# Part III: Visualize the links
```{r,warning=FALSE}
temp <- tempfile()
download.file("ftp://ftp.ensembl.org/pub/release-65/gtf/homo_sapiens/Homo_sapiens.GRCh37.65.gtf.gz", temp)
gene_anno <- rtracklayer::readGFF(temp)
unlink(temp)
# rename some columns to match requirements
gene_anno$chromosome <- paste0("chr", gene_anno$seqid)
gene_anno$gene <- gene_anno$gene_id
gene_anno$transcript <- gene_anno$transcript_id
gene_anno$symbol <- gene_anno$gene_name
Plot_connections(direct.net_result, gene_anno, marker = "PYGB", cutoff = 0.5)
```

# Part IV: Construct regulatory networks
# Identify differential accessiable peaks
```{r, message=FALSE,warning=FALSE}
### identify differential accessiable peaks (DA)
DefaultAssay(combined) <- 'ATAC'
groups <- unique(markers$group)
da_peaks_list <- list()
for (i in 1:length(groups)) {
  print(i)
  da_peaks <- FindMarkers(
    object = combined,
    min.pct = 0.05,
    logfc.threshold = 0.1,
    ident.1 = groups[i],
    group.by = "celltype",
    test.use = 'LR'
  )
  da_peaks_list[[i]] <- da_peaks
}
```
# Detect CRE-TF connections
```{r, message=FALSE,warning=FALSE}
# We have inferred links for all differential genes ( set focus_markers to features.rna)
load("/Users/lihuazhang/Documents/DIRECT-NET/A549/Data/A549_direct.net.RData")
# check the function type name
direct.net_result$function_type <- gsub("HF","HC",direct.net_result$function_type)
direct.net_result$function_type <- gsub("Rest","MC",direct.net_result$function_type)
direct.net_result$function_type <- gsub("LF","LC",direct.net_result$function_type)
# CRE-gene connections
CREs_Gene <- generate_CRE_Gene_links(direct.net_result, markers)
# Find focused CREs which is overlapped with DA
Focused_CREs <- generate_CRE(L_G_record = CREs_Gene$distal, P_L_G_record = CREs_Gene$promoter, da_peaks_list)
# detect TFs for distal CREs
library(BSgenome.Hsapiens.UCSC.hg19)
L_TF_record <- generate_peak_TF_links(peaks_bed_list = Focused_CREs$distal, species="Homo sapiens", genome = BSgenome.Hsapiens.UCSC.hg19, markers)
# detect TFs for Promoters
P_L_TF_record <- generate_peak_TF_links(peaks_bed_list = Focused_CREs$promoter, species="Homo sapiens", genome = BSgenome.Hsapiens.UCSC.hg19, markers)

```

# Output TF-gene connections
```{r, message=FALSE,warning=FALSE}
network_links <- generate_links_for_Cytoscape(L_G_record = CREs_Gene$distal, L_TF_record, P_L_G_record = CREs_Gene$promoter, P_L_TF_record,groups)
```

# Ouput Node attribute
```{r, message=FALSE,warning=FALSE}
Node_attribute <- generate_node_for_Cytoscape(network_links,markers)
```
