---
title: "brain_development"
output: html_document
date: "`r Sys.Date()`"
---

## data information:

Lab: Kriegstein Lab

Institution: University of California, San Francisco

Publication: [Velmeshev et al. Science. 2023.](https://www.science.org/doi/10.1126/science.adf0834)

PubMed Abstract: [37824647](https://pubmed.ncbi.nlm.nih.gov/37824647/)

Publication Fulltext: [10.1126/science.adf0834](https://www.science.org/doi/10.1126/science.adf0834)

Submitted by: Dmitry Velmeshev ([dmitry.velmeshev\@duke.edu](mailto:dmitry.velmeshev@duke.edu){.email}) (2023-08-30), Version 2

Scanpy h5ad: [velmeshev_snRNA_seq.h5ad](https://cells.ucsc.edu/pre-postnatal-cortex/all/rna/velmeshev_snRNA_seq.h5ad)

Direct link to this collection for manuscripts: <https://pre-postnatal-cortex.cells.ucsc.edu>

## data link:

snrna-seq: <https://cells.ucsc.edu/?ds=pre-postnatal-cortex+all+rna>

snatac-seq: <https://cells.ucsc.edu/?ds=pre-postnatal-cortex+all+atac>

gene activity level: <https://cells.ucsc.edu/?ds=pre-postnatal-cortex+all+ga>

h5ad and Seurat (V5) object: <https://cellxgene.cziscience.com/collections/bacccb91-066d-4453-b70e-59de0b4598cd>

## code link

zenodo: <https://zenodo.org/records/7245297>

## run

```{r load_libraries, message=FALSE}
# library(Seurat)
library(magrittr)
library(Seurat)
library(Signac)
library(EnsDb.Hsapiens.v86)
library(ggplot2)
library(cowplot)
```

```{r load_snRNA_data}
# rna_matrix <- Seurat::Read10X(
#   data.dir = "../../inferCSN_data/science_data/rna",
#   gene.column = 1)
# saveRDS(
#   rna_matrix,
#   "../../inferCSN_data/science_data/rna_output/rna_matrix.rds")

rna_matrix <- readRDS("../../inferCSN_data/science_data/rna_output/rna_matrix.rds")

BPCells::write_matrix_dir(
  mat = rna_matrix,
  dir = '../../inferCSN_data/science_data/bpcells/rna')
rna_matrix <- BPCells::open_matrix_dir(dir = "../../inferCSN_data/science_data/bpcells/rna")

rna_metadata <- read.table(
  file = "../../inferCSN_data/science_data/rna_metadata/meta.tsv",
  sep = "\t",
  header = TRUE,
  row.names = 1)
rna_metadata$modality <- "RNA"

rna_umap <- read.table(
  file = "../../inferCSN_data/science_data/rna_metadata/UMAP.coords.tsv.gz",
  sep = "\t",
  row.names = 1)

group <- c(
  "Excitatory neurons", # 0
  "Oligodendrocytes", # 1
  "Interneurons", # 2
  "Excitatory neurons", # 3
  "Astrocytes", # 4
  "Astrocytes", # 5
  "OPCs", # 6
  "Excitatory neurons", # 7
  "Interneurons", # 8
  "Interneurons", # 9
  "Interneurons", # 10
  "Excitatory neurons", # 11
  "Excitatory neurons", # 12
  "Excitatory neurons", # 13
  "Excitatory neurons", # 14
  "Interneurons", # 15
  "Excitatory neurons", # 16
  "Excitatory neurons", # 17
  "Excitatory neurons", # 18
  "Microglia", # 19
  "Excitatory neurons", # 20
  "Excitatory neurons", # 21
  "Excitatory neurons", # 22
  "Excitatory neurons", # 23
  "Excitatory neurons", # 24
  "Glial progenitors", # 25
  "Vascular cells", # 26
  "Interneurons", # 27
  "Excitatory neurons", # 28
  "Excitatory neurons", # 29
  "Excitatory neurons", # 30
  "Interneurons", # 31
  "Excitatory neurons", # 32
  "Excitatory neurons", # 33
  "Excitatory neurons", # 34
  "Unkown", # 35
  "Excitatory neurons", # 36
  "Excitatory neurons", # 37
  "Unkown", # 38
  "Unkown" # 39
  )


rna_metadata$cell_type1 <- plyr::mapvalues(
  rna_metadata$seurat_clusters,
  from = unique(rna_metadata$seurat_clusters),
  to = group)

seurat_object <- Seurat::CreateSeuratObject(
  counts = rna_matrix,
  meta.data = rna_metadata); rm(rna_matrix)


seurat_object <- seurat_object %>%
    Seurat::NormalizeData() %>% 
    Seurat::ScaleData() %>% 
    # Seurat::PercentageFeatureSet(pattern = "^MT-", col.name = "percent.mt") %>%
    # Seurat::SCTransform(vars.to.regress = "percent.mt") %>%
    Seurat::FindVariableFeatures() %>%
    Seurat::RunPCA() %>%
    Seurat::FindNeighbors(dims = 1:30) %>%
    Seurat::RunUMAP(dims = 1:30)

Idents(seurat_object) <- seurat_object$cell_type1

colnames(rna_umap) <- c("UMAP_1", "UMAP_2")
seurat_object@reductions$umap@cell.embeddings <- as.matrix(rna_umap)
gc()
```

```{r download_from_cellxgene(5647MB)_just_rna, message=FALSE}
if (!file.exists("../../inferCSN_data/science_data/seurat_v5.rds")) {
  options(timeout = 3600)
  download.file(
    "https://datasets.cellxgene.cziscience.com/50671879-9cc9-4a91-944a-6dbcf1c67173.rds",
    "../../inferCSN_data/science_data/seurat_v5.rds")
}
seurat_object <- readRDS("../../inferCSN_data/science_data/seurat_v5.rds")
group <- c(
  "OPCs", # 6
  "Vascular cells", # 26
  "Excitatory neurons", # 17
  "Excitatory neurons", # 22
  "Oligodendrocytes", # 1
  "Interneurons", # 2
  "Excitatory neurons", # 14
  "Microglia", # 19
  "Interneurons", # 10
  "Astrocytes", # 5
  "Excitatory neurons", # 24
  "Unkown", # 38
  "Excitatory neurons", # 29
  "Excitatory neurons", # 13
  "Unkown", # 35
  "Astrocytes", # 4
  "Interneurons", # 31
  "Interneurons", # 8
  "Excitatory neurons", # 21
  "Unkown", # 39
  "Interneurons", # 9
  "Excitatory neurons", # 20
  "Excitatory neurons", # 36
  "Excitatory neurons", # 30
  "Excitatory neurons", # 33
  "Excitatory neurons", # 23
  "Excitatory neurons", # 7
  "Excitatory neurons", # 12
  "Excitatory neurons", # 0
  "Excitatory neurons", # 18
  "Excitatory neurons", # 28
  "Interneurons", # 15
  "Interneurons", # 27
  "Excitatory neurons", # 16
  "Glial progenitors", # 25
  "Excitatory neurons", # 11
  "Excitatory neurons", # 3
  "Excitatory neurons", # 32
  "Excitatory neurons", # 34
  "Excitatory neurons"# 37
  )
seurat_object@meta.data <- rna_metadata
Seurat::Idents(seurat_object) <- seurat_object$seurat_clusters
names(group) <- levels(seurat_object)
seurat_object <- Seurat::RenameIdents(seurat_object, group)
seurat_object$celltype <- Seurat::Idents(seurat_object)
```

```{r plot, size=5}
save(seurat_object, file = "../../inferCSN_data/science_data/seurat_rna_annotation.rds")
load("../../inferCSN_data/science_data/seurat_rna_annotation.rds")
p1 <- Seurat::DimPlot(seurat_object, reduction = "umap", group.by = "seurat_clusters", raster = FALSE, label = TRUE) + NoLegend()
p2 <- Seurat::DimPlot(seurat_object, reduction = "umap", group.by = "celltype", raster = FALSE, label = TRUE) + NoLegend()
p1 + p2
```

```{r}
seurat_object <- subset(seurat_object, region == "cortex")
p1 <- Seurat::DimPlot(seurat_object, reduction = "umap", group.by = "seurat_clusters", raster = FALSE, label = TRUE) + NoLegend()
p2 <- Seurat::DimPlot(seurat_object, reduction = "umap", group.by = "celltype", raster = FALSE, label = TRUE) + NoLegend()
p1 + p2
```

```{r save.img, include = TRUE}
plot <- (p1 + p2) &
  xlab("UMAP 1") & ylab("UMAP 2") & 
  theme(axis.title = element_text(size = 10))
ggsave(
  filename = "../../inferCSN_data/science_data/rna-seq_annotation_cortex.jpg",
  height = 5,
  width = 10,
  plot = plot,
  quality = 50)
```

```{r load_snATAC_data}
# atac_matrix <- Seurat::Read10X(
#   data.dir = "../../inferCSN_data/science_data/rna",
#   gene.column = 1)
# saveRDS(
#   atac_matrix,
#   "../../inferCSN_data/science_data/atac_output/atac_matrix.rds")
atac_matrix <- readRDS("../../inferCSN_data/science_data/atac_output/atac_matrix.rds")

atac_metadata <- read.table(
  file = "../../inferCSN_data/science_data/atac_metadata/meta.tsv",
  sep = "\t",
  header = TRUE,
  row.names = 1)

atac_umap <- read.table(
  file = "../../inferCSN_data/science_data/atac_metadata/UMAP.coords.tsv.gz",
  sep = "\t",
  row.names = 1)

atac_metadata$modality <- "ATAC"
atac_metadata <- atac_metadata[atac_metadata$region == "Cortex", ]
atac_umap <- atac_umap[rownames(atac_metadata), ]
atac_matrix <- atac_matrix[, rownames(atac_metadata)]
```

```{r atac, message=FALSE}
# Now add in the ATAC-seq data
# we'll only use peaks in standard chromosomes
genome.info <- read.table(file = paste0("../../inferCSN_data/hg38.promoter.regions.txt"))

names(genome.info) <- c("Chrom","Starts","Ends","genes")
genes <- lapply(genome.info$genes, function(x) strsplit(x,"[|]")[[1]][1])
genes <- lapply(genes, function(x) strsplit(x, "[.]")[[1]][1])
genes <- unlist(genes)
genome.info$genes <- genes
unik <- !duplicated(genes)
genome.info <- genome.info[unik,]

peak <- rownames(atac_matrix)
peak1 <- strsplit(peak, "-")
peaks <- matrix(0, nrow = length(peak), ncol = 1)
for (i in 1:length(peaks)) {
  # peaks[i] <- paste0("chr",peak1[[i]][1],":",peak1[[i]][2],"-",peak1[[i]][3])
  peaks[i] <- paste0(peak1[[i]][1],":",peak1[[i]][2],"-",peak1[[i]][3])
}
rownames(atac_matrix) <- as.vector(peaks)

grange.counts <- Signac::StringToGRanges(rownames(atac_matrix), sep = c(":", "-"))
grange.use <- as.vector(GenomeInfoDb::seqnames(grange.counts)) %in% GenomeInfoDb::standardChromosomes(grange.counts)
atac_matrix <- atac_matrix[as.vector(grange.use), ]
annotations <- Signac::GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86)
GenomeInfoDb::seqlevelsStyle(annotations) <- 'UCSC'
genome.name <- "hg38"
Signac::genome(annotations) <- genome.name
```

```{r}
chrom_assay <- Seurat::CreateSeuratObject(
  counts = atac_matrix,
  meta.data = atac_metadata,
  assay = "ATAC"); rm(atac_matrix)

chrom_assay <- Signac::CreateChromatinAssay(
  counts = atac_matrix,
  sep = c(":", "-"),
  genome = genome.name,
  min.cells = 0,
  annotation = annotations); rm(atac_matrix)


Signac::Annotation(chrom_assay) <- annotations
seurat_object[["ATAC"]] <- chrom_assay

save(chrom_assay, file = "../../inferCSN_data/science_data/chrom_assay.rds")
load("../../inferCSN_data/science_data/chrom_assay.rds")


chrom_assay <- RunTFIDF(chrom_assay)
chrom_assay <- FindTopFeatures(chrom_assay, min.cutoff = "q0")
chrom_assay <- RunSVD(chrom_assay)
chrom_assay <- RunUMAP(chrom_assay, reduction = "lsi", dims = 2:30, reduction.name = "umap.atac", reduction.key = "atacUMAP_")
```
# Identifying anchors between scRNA-seq and scATAC-seq datasets

In order to identify 'anchors' between scRNA-seq and scATAC-seq experiments, we first generate a rough estimate of the transcriptional activity of each gene by quantifying ATAC-seq counts in the 2 kb-upstream region and gene body, using the `GeneActivity()` function in the Signac package. The ensuing gene activity scores from the scATAC-seq data are then used as input for canonical correlation analysis, along with the gene expression quantifications from scRNA-seq. We perform this quantification for all genes identified as being highly variable from the scRNA-seq dataset.
```{r gene.activity}
# quantify gene activity
seurat_object <- Seurat::FindVariableFeatures(seurat_object)
DefaultAssay(chrom_assay) <- "ATAC"
gene.activities <- GeneActivity(chrom_assay, features = VariableFeatures(seurat_object))

gene.activities <- Seurat::Read10X(
  data.dir = "../../inferCSN_data/science_data/gene_activity_level",
  gene.column = 1)

gene.activities_metadata <- read.table(
  file = "../../inferCSN_data/science_data/gene_activity_metadata/meta.tsv",
  sep = "\t",
  header = TRUE,
  row.names = 1)

gene.activities_umap <- read.table(
  file = "../../inferCSN_data/science_data/gene_activity_metadata/UMAP.coords.tsv.gz",
  sep = "\t",
  row.names = 1)
gene.activities <- gene.activities[, rownames(atac_metadata)]

# add gene activities as a new assay
chrom_assay[["ACTIVITY"]] <- Seurat::CreateAssayObject(counts = gene.activities)

# normalize gene activities
Seurat::DefaultAssay(chrom_assay) <- "ACTIVITY"
chrom_assay <- Seurat::NormalizeData(chrom_assay)
chrom_assay <- Seurat::ScaleData(chrom_assay, features = rownames(chrom_assay))
```

```{r label.xfer}
# Identify anchors 
transfer.anchors <- Seurat::FindTransferAnchors(
  reference = seurat_object,
  query = chrom_assay,
  features = Seurat::VariableFeatures(object = seurat_object),
  reference.assay = 'RNA',
  query.assay = 'ACTIVITY',
  reduction = 'cca'
)
```

# Annotate scATAC-seq cells via label transfer

After identifying anchors, we can transfer annotations from the scRNA-seq dataset onto the scATAC-seq cells. The annotations are stored in the `seurat_annotations` field, and are provided as input to the `refdata` parameter. The output will contain a matrix with predictions and confidence scores for each ATAC-seq cell.

```{r transfer.data}
celltype.predictions <- TransferData(
  anchorset = transfer.anchors,
  refdata = seurat_object$seurat_annotations,
  weight.reduction = chrom_assay[['lsi']],
  dims = 2:30
)

chrom_assay <- AddMetaData(chrom_assay, metadata = celltype.predictions)
```
