<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Fit models for gene expression — fit_models • inferCSN</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Fit models for gene expression — fit_models"><meta property="og:description" content="Fit models for gene expression"><meta property="og:image" content="https://mengxu98.github.io/inferCSNplus/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">inferCSN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mengxu98/inferCSNplus" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/mengxu98" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fit models for gene expression</h1>
    <small class="dont-index">Source: <a href="https://github.com/mengxu98/inferCSNplus/blob/HEAD/R/setGenerics.R" class="external-link"><code>R/setGenerics.R</code></a>, <a href="https://github.com/mengxu98/inferCSNplus/blob/HEAD/R/inferCSN-methods.R" class="external-link"><code>R/inferCSN-methods.R</code></a></small>
    <div class="hidden name"><code>fit_models.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Fit models for gene expression</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fit_models</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for CSNObject</span></span>
<span><span class="fu">fit_models</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  network_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">method</span>, <span class="st">"_network"</span><span class="op">)</span>,</span>
<span>  peak_to_gene_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Signac"</span>, <span class="st">"GREAT"</span><span class="op">)</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">1e+05</span>,</span>
<span>  downstream <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  extend <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  only_tss <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  peak_to_gene_domains <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tf_cor <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  peak_cor <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  aggregate_rna_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  aggregate_peaks_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"srm"</span>, <span class="st">"glm"</span>, <span class="st">"glmnet"</span>, <span class="st">"cv.glmnet"</span>, <span class="st">"brms"</span>, <span class="st">"xgb"</span>, <span class="st">"bagging_ridge"</span>,</span>
<span>    <span class="st">"bayesian_ridge"</span><span class="op">)</span>,</span>
<span>  interaction_term <span class="op">=</span> <span class="st">":"</span>,</span>
<span>  adjust_method <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>The input data, a csn object.</p></dd>


<dt>...</dt>
<dd><p>Arguments for other methods</p></dd>


<dt>genes</dt>
<dd><p>A character vector with the target genes to consider for CSN inference.
Takes all <code>VariableFeatures</code> in the object per default.</p></dd>


<dt>network_name</dt>
<dd><p>network_name.</p></dd>


<dt>peak_to_gene_method</dt>
<dd><p>Character specifying the method to
link peak overlapping motif regions to nearby genes. One of <code>Signac</code> or <code>GREAT</code>.</p></dd>


<dt>upstream</dt>
<dd><p>Integer defining the distance upstream of the gene to consider as potential regulatory region.</p></dd>


<dt>downstream</dt>
<dd><p>Integer defining the distance downstream of the gene to consider as potential regulatory region.</p></dd>


<dt>extend</dt>
<dd><p>Integer defining the distance from the upstream and downstream of the basal regulatory region.
Only used of `peak_to_gene_method = 'GREAT'`.</p></dd>


<dt>only_tss</dt>
<dd><p>Logical. Measure distance from the TSS (<code>TRUE</code>) or from the entire gene body (<code>FALSE</code>).</p></dd>


<dt>peak_to_gene_domains</dt>
<dd><p><code>GenomicRanges</code> object with regulatory regions for each gene.</p></dd>


<dt>tf_cor</dt>
<dd><p>Threshold for TF - target gene correlation.</p></dd>


<dt>peak_cor</dt>
<dd><p>Threshold for binding peak - target gene correlation.</p></dd>


<dt>aggregate_rna_col</dt>
<dd><p>aggregate_rna_col</p></dd>


<dt>aggregate_peaks_col</dt>
<dd><p>aggregate_peaks_col</p></dd>


<dt>method</dt>
<dd><p>A character string indicating the method to fit the model.
* <code>'srm'</code> - Sparse Regression Model.
* <code>'glm'</code> - Generalized Liner Model with <code><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></code>.
* <code>'glmnet'</code>, <code>'cv.glmnet'</code> - Regularized Generalized Liner Model with <code>glmnet</code>.
* <code>'brms'</code> - Bayesian Regression Models using <code>brms-package</code>.
* <code>'xgb'</code> - Gradient Boosting Regression using <code>xgboost</code>.
* <code>'bagging_ridge'</code> - Bagging Ridge Regression using scikit-learn via <a href="https://rstudio.github.io/reticulate/reference/reticulate.html" class="external-link">reticulate</a>.
* <code>'bayesian_ridge'</code> - Bayesian Ridge Regression using scikit-learn via <a href="https://rstudio.github.io/reticulate/reference/reticulate.html" class="external-link">reticulate</a>.</p></dd>


<dt>interaction_term</dt>
<dd><p>The interaction term to use in the model between TF and binding site.
* <code>'+'</code> for additive interaction.
* <code>':'</code> for 'multiplicative' interaction.
* <code>'*'</code> for crossing interaction, i.e. additive AND 'multiplicative'.
For more info, see <code><a href="https://rdrr.io/r/stats/formula.html" class="external-link">formula</a></code></p></dd>


<dt>adjust_method</dt>
<dd><p>Method for adjusting p-values.</p></dd>


<dt>scale</dt>
<dd><p>Logical. Whether to z-transform the expression and accessibility matrices.</p></dd>


<dt>verbose</dt>
<dd><p>Logical value. Whether to print detailed information.</p></dd>


<dt>cores</dt>
<dd><p>Number of CPU cores used. Setting to parallelize the computation with <code><a href="https://rdrr.io/pkg/foreach/man/foreach.html" class="external-link">foreach</a></code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A CSNObject.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Meng Xu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

