<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Initiate the CSNObject object — initiate_object • inferCSN</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Initiate the CSNObject object — initiate_object"><meta property="og:description" content="Initiate the CSNObject object"><meta property="og:image" content="https://mengxu98.github.io/inferCSNplus/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">inferCSN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mengxu98/inferCSNplus" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
<li>
  <a href="https://github.com/mengxu98" class="external-link">
    <span class="fas fa-home fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Initiate the <code>CSNObject</code> object</h1>
    <small class="dont-index">Source: <a href="https://github.com/mengxu98/inferCSNplus/blob/main/R/setGenerics.R" class="external-link"><code>R/setGenerics.R</code></a>, <a href="https://github.com/mengxu98/inferCSNplus/blob/main/R/initiate_object.R" class="external-link"><code>R/initiate_object.R</code></a></small>
    <div class="hidden name"><code>initiate_object.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Initiate the <code>CSNObject</code> object</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">initiate_object</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'matrix'</span></span>
<span><span class="fu">initiate_object</span><span class="op">(</span><span class="va">object</span>, regulators <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, targets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'Seurat'</span></span>
<span><span class="fu">initiate_object</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  celltype <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  celltype_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter_mode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"variable"</span>, <span class="st">"celltype"</span>, <span class="st">"unfiltered"</span><span class="op">)</span>,</span>
<span>  filter_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"aggregate"</span>, <span class="st">"celltype"</span><span class="op">)</span>,</span>
<span>  rna_assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  rna_min_pct <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  rna_logfc_threshold <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  rna_test_method <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>  n_variable_genes <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  peak_assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  peak_min_pct <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  peak_logfc_threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  peak_test_method <span class="op">=</span> <span class="st">"LR"</span>,</span>
<span>  n_variable_peaks <span class="op">=</span> <span class="st">"q5"</span>,</span>
<span>  regions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exclude_exons <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  only_pos <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  p_value <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'CSNObject'</span></span>
<span><span class="fu">initiate_object</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>A Seurat object containing gene expression and/or chromatin accessibility data.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to marker detection functions.</p></dd>


<dt id="arg-regulators">regulators<a class="anchor" aria-label="anchor" href="#arg-regulators"></a></dt>
<dd><p>Regulators to consider for CSN inference.</p></dd>


<dt id="arg-targets">targets<a class="anchor" aria-label="anchor" href="#arg-targets"></a></dt>
<dd><p>Targets to consider for CSN inference.</p></dd>


<dt id="arg-celltype">celltype<a class="anchor" aria-label="anchor" href="#arg-celltype"></a></dt>
<dd><p>Selected celltype(s) to analyze. If NULL, all cells will be used.</p></dd>


<dt id="arg-celltype-by">celltype_by<a class="anchor" aria-label="anchor" href="#arg-celltype-by"></a></dt>
<dd><p>Column name in Seurat meta.data for cell type information.
If NULL, all cells will be treated as one group.</p></dd>


<dt id="arg-filter-mode">filter_mode<a class="anchor" aria-label="anchor" href="#arg-filter-mode"></a></dt>
<dd><p>Filter mode for identifying cell-type specific features.</p></dd>


<dt id="arg-filter-by">filter_by<a class="anchor" aria-label="anchor" href="#arg-filter-by"></a></dt>
<dd><p>Character. When filter_mode is "variable",
specify whether to calculate variable features using "aggregate" cells or by "celltype".
Default is "aggregate".</p></dd>


<dt id="arg-rna-assay">rna_assay<a class="anchor" aria-label="anchor" href="#arg-rna-assay"></a></dt>
<dd><p>A character vector indicating the name of the gene expression
assay in the <code>Seurat</code> object.</p></dd>


<dt id="arg-rna-min-pct">rna_min_pct<a class="anchor" aria-label="anchor" href="#arg-rna-min-pct"></a></dt>
<dd><p>Minimum percentage of cells expressing the gene in either population.</p></dd>


<dt id="arg-rna-logfc-threshold">rna_logfc_threshold<a class="anchor" aria-label="anchor" href="#arg-rna-logfc-threshold"></a></dt>
<dd><p>Minimum log fold-change threshold for RNA data.</p></dd>


<dt id="arg-rna-test-method">rna_test_method<a class="anchor" aria-label="anchor" href="#arg-rna-test-method"></a></dt>
<dd><p>Statistical test to use for gene markers identification.</p></dd>


<dt id="arg-n-variable-genes">n_variable_genes<a class="anchor" aria-label="anchor" href="#arg-n-variable-genes"></a></dt>
<dd><p>Number of variable features to select when not comparing between groups.</p></dd>


<dt id="arg-peak-assay">peak_assay<a class="anchor" aria-label="anchor" href="#arg-peak-assay"></a></dt>
<dd><p>A character vector indicating the name of the chromatin
accessibility assay in the <code>Seurat</code> object.
If NULL, only RNA-seq data will be processed.</p></dd>


<dt id="arg-peak-min-pct">peak_min_pct<a class="anchor" aria-label="anchor" href="#arg-peak-min-pct"></a></dt>
<dd><p>Minimum percentage of cells expressing the peak in either population.</p></dd>


<dt id="arg-peak-logfc-threshold">peak_logfc_threshold<a class="anchor" aria-label="anchor" href="#arg-peak-logfc-threshold"></a></dt>
<dd><p>Minimum log fold-change threshold for peak data.</p></dd>


<dt id="arg-peak-test-method">peak_test_method<a class="anchor" aria-label="anchor" href="#arg-peak-test-method"></a></dt>
<dd><p>Statistical test to use for peak markers identification.</p></dd>


<dt id="arg-n-variable-peaks">n_variable_peaks<a class="anchor" aria-label="anchor" href="#arg-n-variable-peaks"></a></dt>
<dd><p>Number of variable peaks to select ("q5" by default).</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>Candidate regions to consider for binding site inference.
If <code>NULL</code>, all peaks regions are considered.
<code>regions</code> could be a <code>GRanges</code> object,
or a data frame with three columns: chrom, start, end.</p></dd>


<dt id="arg-exclude-exons">exclude_exons<a class="anchor" aria-label="anchor" href="#arg-exclude-exons"></a></dt>
<dd><p>Logical. Whether to consider exons for binding site inference.</p></dd>


<dt id="arg-only-pos">only_pos<a class="anchor" aria-label="anchor" href="#arg-only-pos"></a></dt>
<dd><p>Only return positive markers.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Print progress messages.</p></dd>


<dt id="arg-p-value">p_value<a class="anchor" aria-label="anchor" href="#arg-p-value"></a></dt>
<dd><p>Significance level threshold for filtering features (default: 0.05).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>CSNObject object.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Meng Xu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

      </footer></div>






  </body></html>

