<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ inferCSN</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://mengxu98.github.io/inferCSNplus/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">inferCSN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mengxu98/inferCSNplus" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/mengxu98" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="inferring-gene-regulatory-network">Inferring gene regulatory network <a href="#inferring-gene-regulatory-network" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="infercsn">inferCSN <a href="#infercsn" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="inferCSN-package.html">inferCSN-package</a></code> </p>
        </td>
        <td><p>Inferring cell-type-specific gene regulatory networks from single-cell omics data</p></td>
      </tr><tr><td>
          <p><code><a href="inferCSN.logo.html">inferCSN.logo()</a></code> </p>
        </td>
        <td><p>inferCSN logo</p></td>
      </tr><tr><td>
          <p><code><a href="inferCSN.html">inferCSN()</a></code> </p>
        </td>
        <td><p>Inferring Cell-Specific Gene Regulatory Network</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="sparse-regression-model">Sparse regression model <a href="#sparse-regression-model" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="fit_models.html">fit_models()</a></code> </p>
        </td>
        <td><p>Fit models for gene expression</p></td>
      </tr><tr><td>
          <p><code><a href="fit_model.html">fit_model()</a></code> </p>
        </td>
        <td><p>Fit (regularized) generalized linear model</p></td>
      </tr><tr><td>
          <p><code><a href="format_coefs.html">format_coefs()</a></code> </p>
        </td>
        <td><p>Format network coefficients</p></td>
      </tr><tr><td>
          <p><code><a href="model.fit.html">model.fit()</a></code> </p>
        </td>
        <td><p>Fit a sparse regression model</p></td>
      </tr><tr><td>
          <p><code><a href="single.network.html">single.network()</a></code> </p>
        </td>
        <td><p>Construct network for single gene</p></td>
      </tr><tr><td>
          <p><code><a href="sparse.regression.html">sparse.regression()</a></code> </p>
        </td>
        <td><p>Sparse regression model</p></td>
      </tr><tr><td>
          <p><code><a href="fit_bagging_ridge.html">fit_bagging_ridge()</a></code> </p>
        </td>
        <td><p>Fit a bagging ridge regression model as implemented in scikit-learn (python)</p></td>
      </tr><tr><td>
          <p><code><a href="fit_bayesian_ridge.html">fit_bayesian_ridge()</a></code> </p>
        </td>
        <td><p>Fit a bayesian ridge regression model as implemented in scikit-learn (python)</p></td>
      </tr><tr><td>
          <p><code><a href="fit_brms.html">fit_brms()</a></code> </p>
        </td>
        <td><p>Fit a Bayesian regression model with brms and Stan</p></td>
      </tr><tr><td>
          <p><code><a href="fit_cvglmnet.html">fit_cvglmnet()</a></code> </p>
        </td>
        <td><p>Cross-validation for regularized generalized linear models</p></td>
      </tr><tr><td>
          <p><code><a href="fit_cvsrm.html">fit_cvsrm()</a></code> </p>
        </td>
        <td><p>Fit a sparse regression model</p></td>
      </tr><tr><td>
          <p><code><a href="fit_glm.html">fit_glm()</a></code> </p>
        </td>
        <td><p>Fit generalized linear model</p></td>
      </tr><tr><td>
          <p><code><a href="fit_glmnet.html">fit_glmnet()</a></code> </p>
        </td>
        <td><p>Fit regularized generalized linear model</p></td>
      </tr><tr><td>
          <p><code><a href="fit_srm.html">fit_srm()</a></code> </p>
        </td>
        <td><p>Fit a sparse regression model</p></td>
      </tr><tr><td>
          <p><code><a href="fit_xgb.html">fit_xgb()</a></code> </p>
        </td>
        <td><p>Fit a gradient boosting regression model with XGBoost</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="class">Class <a href="#class" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="initiate_object.html">initiate_object()</a></code> </p>
        </td>
        <td><p>Initiate the <code>RegulatoryNetwork</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="CSNObject-class.html">CSNObject-class</a></code> <code><a href="CSNObject-class.html">CSNObject</a></code> </p>
        </td>
        <td><p>The CSNObject class</p></td>
      </tr><tr><td>
          <p><code><a href="CSNObjectList-class.html">CSNObjectList-class</a></code> <code><a href="CSNObjectList-class.html">CSNObjectList</a></code> </p>
        </td>
        <td><p>The CSNObjectList class</p></td>
      </tr><tr><td>
          <p><code><a href="Modules-class.html">Modules-class</a></code> <code><a href="Modules-class.html">Modules</a></code> </p>
        </td>
        <td><p>The Modules class</p></td>
      </tr><tr><td>
          <p><code><a href="Network-class.html">Network-class</a></code> <code><a href="Network-class.html">Network</a></code> </p>
        </td>
        <td><p>The Network class</p></td>
      </tr><tr><td>
          <p><code><a href="Regions-class.html">Regions-class</a></code> <code><a href="Regions-class.html">Regions</a></code> </p>
        </td>
        <td><p>The Regions class</p></td>
      </tr><tr><td>
          <p><code><a href="RegulatoryNetwork-class.html">RegulatoryNetwork-class</a></code> <code><a href="RegulatoryNetwork-class.html">RegulatoryNetwork</a></code> </p>
        </td>
        <td><p>The RegularotyNetwork class</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="vector-functions">vector functions <a href="#vector-functions" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="infer_vector.html">infer_vector()</a></code> </p>
        </td>
        <td><p>infer VECTOR</p></td>
      </tr><tr><td>
          <p><code><a href="vector_auto_center.html">vector_auto_center()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="vector_build_grid.html">vector_build_grid()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="vector_build_net.html">vector_build_net()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="vector_build_value.html">vector_build_value()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="vector_draw_arrow.html">vector_draw_arrow()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="vector_grid_value.html">vector_grid_value()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="vector_rank_pca.html">vector_rank_pca()</a></code> </p>
        </td>
        <td><p>rank umap</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="rna-data-processing">RNA data processing <a href="#rna-data-processing" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="normalize.seurat.object.html">normalize.seurat.object()</a></code> </p>
        </td>
        <td><p>normalize.seurat.object</p></td>
      </tr><tr><td>
          <p><code><a href="generate_aggregated_data.html">generate_aggregated_data()</a></code> </p>
        </td>
        <td><p>Create aggregated data for a certain cluster</p></td>
      </tr><tr><td>
          <p><code><a href="aggregating.data.html">aggregating.data()</a></code> </p>
        </td>
        <td><p>Create aggregated data for inferCSN</p></td>
      </tr><tr><td>
          <p><code><a href="as.slam.matrix.html">as.slam.matrix()</a></code> </p>
        </td>
        <td><p>Convert a sparseMatrix from Matrix package to a slam matrix</p></td>
      </tr><tr><td>
          <p><code><a href="as_sparse.html">as_sparse()</a></code> </p>
        </td>
        <td><p>Convert a slam matrix to a sparseMatrix</p></td>
      </tr><tr><td>
          <p><code><a href="estimate.size.factors.html">estimate.size.factors()</a></code> </p>
        </td>
        <td><p>Function to calculate the size factor for the single-cell data</p></td>
      </tr><tr><td>
          <p><code><a href="estimate.size.factors.dense.html">estimate.size.factors.dense()</a></code> </p>
        </td>
        <td><p>Estimate size factors dense matrix</p></td>
      </tr><tr><td>
          <p><code><a href="estimate.size.factors.sparse.html">estimate.size.factors.sparse()</a></code> </p>
        </td>
        <td><p>Estimate size factors for each column, given a sparseMatrix from the Matrix package</p></td>
      </tr><tr><td>
          <p><code><a href="estimate.t.html">estimate.t()</a></code> </p>
        </td>
        <td><p>Find the most commonly occuring relative expression value in each cell</p></td>
      </tr><tr><td>
          <p><code><a href="aggregate_assay.html">aggregate_assay()</a></code> </p>
        </td>
        <td><p>Aggregate Seurat assay over groups</p></td>
      </tr><tr><td>
          <p><code><a href="aggregate_matrix.html">aggregate_matrix()</a></code> </p>
        </td>
        <td><p>Aggregate matrix over groups</p></td>
      </tr><tr><td>
          <p><code><a href="expression_ksmooth.html">expression_ksmooth()</a></code> </p>
        </td>
        <td><p>expression_ksmooth</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="peak-data-processing">Peak data processing <a href="#peak-data-processing" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="extract_cre.html">extract_cre()</a></code> </p>
        </td>
        <td><p>extract cre of markers</p></td>
      </tr><tr><td>
          <p><code><a href="extract_cre.gene.links.html">extract_cre.gene.links()</a></code> </p>
        </td>
        <td><p>Extract cre-gene relations of markers</p></td>
      </tr><tr><td>
          <p><code><a href="extract_links.html">extract_links()</a></code> </p>
        </td>
        <td><p>extract links</p></td>
      </tr><tr><td>
          <p><code><a href="extract_network.html">extract_network()</a></code> </p>
        </td>
        <td><p>extract_network</p></td>
      </tr><tr><td>
          <p><code><a href="extract_nodes.html">extract_nodes()</a></code> </p>
        </td>
        <td><p>extract Node character</p></td>
      </tr><tr><td>
          <p><code><a href="extract_peak.tf.links.html">extract_peak.tf.links()</a></code> </p>
        </td>
        <td><p>Identify TFs enriched in cre of focus markers</p></td>
      </tr><tr><td>
          <p><code><a href="find.overlapping.coordinates.html">find.overlapping.coordinates()</a></code> </p>
        </td>
        <td><p>Find peaks that overlap a specific genomic location</p></td>
      </tr><tr><td>
          <p><code><a href="find_peaks_near_genes.html">find_peaks_near_genes()</a></code> </p>
        </td>
        <td><p>Find peaks or regions near gene body or TSS</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="dynamic-object">Dynamic object <a href="#dynamic-object" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="dynamic_genes.html">dynamic_genes()</a></code> </p>
        </td>
        <td><p>Get dynamic genes</p></td>
      </tr><tr><td>
          <p><code><a href="dynamic_genes_new.html">dynamic_genes_new()</a></code> </p>
        </td>
        <td><p>find genes expressed dynamically</p></td>
      </tr><tr><td>
          <p><code><a href="findDynGenes.html">findDynGenes()</a></code> </p>
        </td>
        <td><p>find genes expressed dynamically</p></td>
      </tr><tr><td>
          <p><code><a href="get_pseudotime.html">get_pseudotime()</a></code> </p>
        </td>
        <td><p>Get pseudotime information</p></td>
      </tr><tr><td>
          <p><code><a href="get_embedding.html">get_embedding()</a></code> </p>
        </td>
        <td><p>Get dimensional information</p></td>
      </tr><tr><td>
          <p><code><a href="compileDynGenes.html">compileDynGenes()</a></code> </p>
        </td>
        <td><p>compileDynGenes</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="network-processing">Network processing <a href="#network-processing" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="network_sift.html">network_sift()</a></code> </p>
        </td>
        <td><p>network_sift</p></td>
      </tr><tr><td>
          <p><code><a href="add_interaction_type.html">add_interaction_type()</a></code> </p>
        </td>
        <td><p>add_interaction_type</p></td>
      </tr><tr><td>
          <p><code><a href="add_type.html">add_type()</a></code> </p>
        </td>
        <td><p>Adds interaction type to dynamic differential network</p></td>
      </tr><tr><td>
          <p><code><a href="DefaultNetwork.html">DefaultNetwork()</a></code> </p>
        </td>
        <td><p>Get active network</p></td>
      </tr><tr><td>
          <p><code><a href="GetAssay.html">GetAssay()</a></code> </p>
        </td>
        <td><p>Get Seurat assay from CSNObject</p></td>
      </tr><tr><td>
          <p><code><a href="GetAssaySummary.html">GetAssaySummary()</a></code> </p>
        </td>
        <td><p>Get summary of seurat assay</p></td>
      </tr><tr><td>
          <p><code><a href="GetGRN.html">GetGRN()</a></code> </p>
        </td>
        <td><p>Get network data</p></td>
      </tr><tr><td>
          <p><code><a href="GetNetwork.html">GetNetwork()</a></code> </p>
        </td>
        <td><p>Get network</p></td>
      </tr><tr><td>
          <p><code><a href="LayerData.html">LayerData()</a></code> </p>
        </td>
        <td><p>Get layer data from CSNObject</p></td>
      </tr><tr><td>
          <p><code><a href="NetworkFeatures.html">NetworkFeatures()</a></code> </p>
        </td>
        <td><p>Get network features</p></td>
      </tr><tr><td>
          <p><code><a href="NetworkGraph.html">NetworkGraph()</a></code> </p>
        </td>
        <td><p>Get network parameters</p></td>
      </tr><tr><td>
          <p><code><a href="NetworkModules.html">NetworkModules()</a></code> </p>
        </td>
        <td><p>Get TF modules</p></td>
      </tr><tr><td>
          <p><code><a href="NetworkParams.html">NetworkParams()</a></code> </p>
        </td>
        <td><p>Get network parameters</p></td>
      </tr><tr><td>
          <p><code><a href="NetworkRegions.html">NetworkRegions()</a></code> </p>
        </td>
        <td><p>Get network regions</p></td>
      </tr><tr><td>
          <p><code><a href="NetworkTFs.html">NetworkTFs()</a></code> </p>
        </td>
        <td><p>Get network TFs</p></td>
      </tr><tr><td>
          <p><code><a href="Params.html">Params()</a></code> </p>
        </td>
        <td><p>Get GRN inference parameters</p></td>
      </tr><tr><td>
          <p><code><a href="VariableFeatures.html">VariableFeatures()</a></code> </p>
        </td>
        <td><p>Get variable features from CSNObject</p></td>
      </tr><tr><td>
          <p><code><a href="crossweight.html">crossweight()</a></code> </p>
        </td>
        <td><p>Perform crossweighting</p></td>
      </tr><tr><td>
          <p><code><a href="define_epochs.html">define_epochs()</a></code> </p>
        </td>
        <td><p>Define epochs</p></td>
      </tr><tr><td>
          <p><code><a href="define_epochs_new.html">define_epochs_new()</a></code> </p>
        </td>
        <td><p>Define epochs</p></td>
      </tr><tr><td>
          <p><code><a href="density_points.html">density_points()</a></code> </p>
        </td>
        <td><p>density_points</p></td>
      </tr><tr><td>
          <p><code><a href="diffnet_community_detection.html">diffnet_community_detection()</a></code> </p>
        </td>
        <td><p>Perform community detection on a dynamic network</p></td>
      </tr><tr><td>
          <p><code><a href="dynamic_windowing.html">dynamic_windowing()</a></code> </p>
        </td>
        <td><p>dynamic_windowing</p></td>
      </tr><tr><td>
          <p><code><a href="dynamic_difference_network.html">dynamic_difference_network()</a></code> </p>
        </td>
        <td><p>Compute a dynamic difference network</p></td>
      </tr><tr><td>
          <p><code><a href="dynamic_shortest_path.html">dynamic_shortest_path()</a></code> </p>
        </td>
        <td><p>Function to return shortest path from 1 regulator to 1 target in a dynamic network</p></td>
      </tr><tr><td>
          <p><code><a href="dynamic_shortest_path_multiple.html">dynamic_shortest_path_multiple()</a></code> </p>
        </td>
        <td><p>Function to return shortest path from multiple TFs to multiple targets in a dynamic network</p></td>
      </tr><tr><td>
          <p><code><a href="edge_uniqueness.html">edge_uniqueness()</a></code> </p>
        </td>
        <td><p>Function to compute edge differences between networks</p></td>
      </tr><tr><td>
          <p><code><a href="epochGRN.html">epochGRN()</a></code> </p>
        </td>
        <td><p>Divides grnDF into epochs, filters interactions between genes not in same or consecutive epochs</p></td>
      </tr><tr><td>
          <p><code><a href="compile_epochs.html">compile_epochs()</a></code> </p>
        </td>
        <td><p>compile_epochs</p></td>
      </tr><tr><td>
          <p><code><a href="get_network_graph.html">get_network_graph()</a></code> </p>
        </td>
        <td><p>Compute network graph embedding using UMAP.</p></td>
      </tr><tr><td>
          <p><code><a href="get_tf_network.html">get_tf_network()</a></code> </p>
        </td>
        <td><p>Get sub-network centered around one TF.</p></td>
      </tr><tr><td>
          <p><code><a href="gof.html">gof()</a></code> </p>
        </td>
        <td><p>Get goodness-of-fit info</p></td>
      </tr><tr><td>
          <p><code><a href="assign_epochs_simple.html">assign_epochs_simple()</a></code> </p>
        </td>
        <td><p>Assigns genes to epochs just based on which mean is maximal</p></td>
      </tr><tr><td>
          <p><code><a href="assign_epochs.html">assign_epochs()</a></code> </p>
        </td>
        <td><p>Assigns genes to epochs</p></td>
      </tr><tr><td>
          <p><code><a href="assign_epochs_new1.html">assign_epochs_new1()</a></code> </p>
        </td>
        <td><p>Assigns genes to epochs</p></td>
      </tr><tr><td>
          <p><code><a href="assign_network.html">assign_network()</a></code> </p>
        </td>
        <td><p>Assigns genes to epochs</p></td>
      </tr><tr><td>
          <p><code><a href="subnets.html">subnets()</a></code> </p>
        </td>
        <td><p>Function to assign nodes to communities via Louvain clustering</p></td>
      </tr><tr><td>
          <p><code><a href="split_epochs_by_group.html">split_epochs_by_group()</a></code> </p>
        </td>
        <td><p>Splits data into epochs manually</p></td>
      </tr><tr><td>
          <p><code><a href="split_epochs_by_pseudotime.html">split_epochs_by_pseudotime()</a></code> </p>
        </td>
        <td><p>Splits data into epochs</p></td>
      </tr><tr><td>
          <p><code><a href="biglist_compute_betweenness_degree.html">biglist_compute_betweenness_degree()</a></code> </p>
        </td>
        <td><p>Computes betweenness and degree of each regulator for each network in a list of networks</p></td>
      </tr><tr><td>
          <p><code><a href="caoGenes.html">caoGenes()</a></code> </p>
        </td>
        <td><p>cluster and order genes</p></td>
      </tr><tr><td>
          <p><code><a href="compute_JI_topregs.html">compute_JI_topregs()</a></code> </p>
        </td>
        <td><p>Computes Jaccard similarity between top regulators in two sets of networks</p></td>
      </tr><tr><td>
          <p><code><a href="compute_betweenness_degree.html">compute_betweenness_degree()</a></code> </p>
        </td>
        <td><p>Function to compute betweenness and degree</p></td>
      </tr><tr><td>
          <p><code><a href="compute_frobenius_distance.html">compute_frobenius_distance()</a></code> </p>
        </td>
        <td><p>Computes frobenius distance in a pairwise manner between two sets of networks</p></td>
      </tr><tr><td>
          <p><code><a href="cor_and_add_action.html">cor_and_add_action()</a></code> </p>
        </td>
        <td><p>Adds an extra column to the result of dynamic_shortest_path_multiple that predicts overall action based on correlation between "from" and "to"</p></td>
      </tr><tr><td>
          <p><code><a href="find_communities.html">find_communities()</a></code> </p>
        </td>
        <td><p>find_communities</p></td>
      </tr><tr><td>
          <p><code><a href="find_cuts_by_clustering.html">find_cuts_by_clustering()</a></code> </p>
        </td>
        <td><p>Returns cuts to define epochs</p></td>
      </tr><tr><td>
          <p><code><a href="find_cuts_by_similarity.html">find_cuts_by_similarity()</a></code> </p>
        </td>
        <td><p>find_cuts_by_similarity</p></td>
      </tr><tr><td>
          <p><code><a href="find_modules.html">find_modules()</a></code> </p>
        </td>
        <td><p>Find TF modules in regulatory network</p></td>
      </tr><tr><td>
          <p><code><a href="find_motifs.html">find_motifs()</a></code> </p>
        </td>
        <td><p>Scan for motifs in candidate regions</p></td>
      </tr><tr><td>
          <p><code><a href="find_targets.html">find_targets()</a></code> </p>
        </td>
        <td><p>Finds binding targets given list of dataframes containing binding info for effectors</p></td>
      </tr><tr><td>
          <p><code><a href="static_shortest_path.html">static_shortest_path()</a></code> </p>
        </td>
        <td><p>Function to return shortest path from 1 regulator to 1 target in a static network</p></td>
      </tr><tr><td>
          <p><code><a href="scale_matrix.html">scale_matrix()</a></code> </p>
        </td>
        <td><p>Standardize the values by rows</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="network-visualization">Network visualization <a href="#network-visualization" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plot_dynamic_networks.html">plot_dynamic_networks()</a></code> </p>
        </td>
        <td><p>plot_dynamic_networks</p></td>
      </tr><tr><td>
          <p><code><a href="plot_static_networks.html">plot_static_networks()</a></code> </p>
        </td>
        <td><p>Plot of dynamic networks</p></td>
      </tr><tr><td>
          <p><code><a href="plot_contrast_networks.html">plot_contrast_networks()</a></code> </p>
        </td>
        <td><p>plot_contrast_networks</p></td>
      </tr><tr><td>
          <p><code><a href="network.heatmap.html">network.heatmap()</a></code> </p>
        </td>
        <td><p>The heatmap of network</p></td>
      </tr><tr><td>
          <p><code><a href="plot_detail_network.html">plot_detail_network()</a></code> </p>
        </td>
        <td><p>Plot the dynamic differential network but colored by communities and optionally faded by igraph::betweenness</p></td>
      </tr><tr><td>
          <p><code><a href="plot_diffnet_detail.html">plot_diffnet_detail()</a></code> </p>
        </td>
        <td><p>Plot the dynamic differential network but colored by communities and optionally faded by betweenness</p></td>
      </tr><tr><td>
          <p><code><a href="plot_dyn_diffnet.html">plot_dyn_diffnet()</a></code> </p>
        </td>
        <td><p>Plot the dynamic differential network</p></td>
      </tr><tr><td>
          <p><code><a href="plot_dynamic_network.html">plot_dynamic_network()</a></code> </p>
        </td>
        <td><p>quick plot of dynamic networks</p></td>
      </tr><tr><td>
          <p><code><a href="plot_gof.html">plot_gof()</a></code> </p>
        </td>
        <td><p>Plot goodness-of-fit metrics.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_heatmap_by_treatment.html">plot_heatmap_by_treatment()</a></code> </p>
        </td>
        <td><p>Useful plotting function to plot heatmap with pre-split matrix</p></td>
      </tr><tr><td>
          <p><code><a href="plot_module_metrics.html">plot_module_metrics()</a></code> </p>
        </td>
        <td><p>Plot module metrics number of genes, number of peaks and number of TFs per gene.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_network_graph.html">plot_network_graph()</a></code> </p>
        </td>
        <td><p>Plot network graph.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_targets_and_regulators.html">plot_targets_and_regulators()</a></code> </p>
        </td>
        <td><p>Updated plot of top regulators given targets in dynamic networks based on a weight column. Top regulators computed for each epoch, but maintained in plot across epochs if present in epoch subnetwork.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_targets_with_top_regulators_detail.html">plot_targets_with_top_regulators_detail()</a></code> </p>
        </td>
        <td><p>quick plot of top regulators given targets in dynamic networks based on reconstruction weight, colored by expression and interaction type</p></td>
      </tr><tr><td>
          <p><code><a href="plot_tf_network.html">plot_tf_network()</a></code> </p>
        </td>
        <td><p>Plot sub-network centered around one TF.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_top_features.html">plot_top_features()</a></code> </p>
        </td>
        <td><p>quick plot of top regulators in dynamic networks</p></td>
      </tr><tr><td>
          <p><code><a href="heatmap_by_treatment_group.html">heatmap_by_treatment_group()</a></code> </p>
        </td>
        <td><p>Useful plotting function to plot heatmap of module expression across time with pre-split matrix</p></td>
      </tr><tr><td>
          <p><code><a href="hm_dyn.html">hm_dyn()</a></code> </p>
        </td>
        <td><p>plots results of findDynGenes</p></td>
      </tr><tr><td>
          <p><code><a href="hm_dyn_clust.html">hm_dyn_clust()</a></code> </p>
        </td>
        <td><p>plots results of findDynGenes</p></td>
      </tr><tr><td>
          <p><code><a href="hm_dyn_epoch.html">hm_dyn_epoch()</a></code> </p>
        </td>
        <td><p>heatmap</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="plotting-functions">Plotting functions <a href="#plotting-functions" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plot_scatter.html">plot_scatter()</a></code> </p>
        </td>
        <td><p>plot_scatter</p></td>
      </tr><tr><td>
          <p><code><a href="plot_embedding.html">plot_embedding()</a></code> </p>
        </td>
        <td><p>plot_embedding</p></td>
      </tr><tr><td>
          <p><code><a href="plot_weight_distribution.html">plot_weight_distribution()</a></code> </p>
        </td>
        <td><p>plot_weight_distribution</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="network-evaluation">Network evaluation <a href="#network-evaluation" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calculate_auc.html">calculate_auc()</a></code> </p>
        </td>
        <td><p>Calculate auc</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_acc.html">calculate_acc()</a></code> </p>
        </td>
        <td><p>Calculate acc</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_network_data.html">prepare_network_data()</a></code> </p>
        </td>
        <td><p>Prepare network data</p></td>
      </tr><tr><td>
          <p><code><a href="JI_across_topregs.html">JI_across_topregs()</a></code> </p>
        </td>
        <td><p>Computes Jaccard similarity between top regulators in two sets of networks across a range of top X regulators</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="calculate-gene-rank">Calculate gene rank <a href="#calculate-gene-rank" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calculate_gene_rank.html">calculate_gene_rank()</a></code> </p>
        </td>
        <td><p>Rank TFs and genes in network</p></td>
      </tr><tr><td>
          <p><code><a href="compute_pagerank.html">compute_pagerank()</a></code> </p>
        </td>
        <td><p>Function to compute page rank of TF+target networks</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="utils-functions">Utils functions <a href="#utils-functions" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="as_matrix.html">as_matrix()</a></code> </p>
        </td>
        <td><p>Convert dgCMatrix into a dense matrix</p></td>
      </tr><tr><td>
          <p><code><a href="check.parameters.html">check.parameters()</a></code> </p>
        </td>
        <td><p>Check input parameters</p></td>
      </tr><tr><td>
          <p><code><a href="coef.SRM_fit.html">coef(<i>&lt;SRM_fit&gt;</i>)</a></code> <code><a href="coef.SRM_fit.html">coef(<i>&lt;SRM_fit_CV&gt;</i>)</a></code> </p>
        </td>
        <td><p>Extracts a specific solution in the regularization path</p></td>
      </tr><tr><td>
          <p><code><a href="coef.html">coef(<i>&lt;CSNObject&gt;</i>)</a></code> <code><a href="coef.html">coef(<i>&lt;RegulatoryNetwork&gt;</i>)</a></code> <code><a href="coef.html">coef(<i>&lt;Network&gt;</i>)</a></code> </p>
        </td>
        <td><p>Get fitted coefficients</p></td>
      </tr><tr><td>
          <p><code><a href="crossweight.html">crossweight()</a></code> </p>
        </td>
        <td><p>Perform crossweighting</p></td>
      </tr><tr><td>
          <p><code><a href="network_format.html">network_format()</a></code> </p>
        </td>
        <td><p>Format weight table</p></td>
      </tr><tr><td>
          <p><code><a href="normalization.html">normalization()</a></code> </p>
        </td>
        <td><p>normalization</p></td>
      </tr><tr><td>
          <p><code><a href="table.to.matrix.html">table.to.matrix()</a></code> </p>
        </td>
        <td><p>Switch weight table to matrix</p></td>
      </tr><tr><td>
          <p><code><a href="filter_sort_matrix.html">filter_sort_matrix()</a></code> </p>
        </td>
        <td><p>Filter and sort matrix</p></td>
      </tr><tr><td>
          <p><code><a href="print.logo.html">print(<i>&lt;logo&gt;</i>)</a></code> </p>
        </td>
        <td><p>print logo figure</p></td>
      </tr><tr><td>
          <p><code><a href="print.SRM_fit.html">print(<i>&lt;SRM_fit&gt;</i>)</a></code> <code><a href="print.SRM_fit.html">print(<i>&lt;SRM_fit_CV&gt;</i>)</a></code> </p>
        </td>
        <td><p>Prints a summary of model.fit</p></td>
      </tr><tr><td>
          <p><code><a href="print.html">print(<i>&lt;RegulatoryNetwork&gt;</i>)</a></code> <code><a href="print.html">print(<i>&lt;Network&gt;</i>)</a></code> <code><a href="print.html">print(<i>&lt;Modules&gt;</i>)</a></code> <code><a href="print.html">print(<i>&lt;Regions&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print RegulatoryNetwork objects</p></td>
      </tr><tr><td>
          <p><code><a href="predict.SRM_fit.html">predict(<i>&lt;SRM_fit&gt;</i>)</a></code> <code><a href="predict.SRM_fit.html">predict(<i>&lt;SRM_fit_CV&gt;</i>)</a></code> </p>
        </td>
        <td><p>Predict Response</p></td>
      </tr><tr><td>
          <p><code><a href="peaks.split.html">peaks.split()</a></code> </p>
        </td>
        <td><p>peaks.split</p></td>
      </tr><tr><td>
          <p><code><a href="ranges.for.coords.html">ranges.for.coords()</a></code> </p>
        </td>
        <td><p>Construct GRanges objects from coordinate strings</p></td>
      </tr><tr><td>
          <p><code><a href="get.tf.gene.html">get.tf.gene()</a></code> </p>
        </td>
        <td><p>Obtain TF and Gene relationships</p></td>
      </tr><tr><td>
          <p><code><a href="transcript.mode.html">transcript.mode()</a></code> </p>
        </td>
        <td><p>use gaussian kernel to calculate the mode of transcript counts</p></td>
      </tr><tr><td>
          <p><code><a href="is.sparse.html">is.sparse()</a></code> </p>
        </td>
        <td><p>is.sparse</p></td>
      </tr><tr><td>
          <p><code><a href="parallelize_fun.html">parallelize_fun()</a></code> </p>
        </td>
        <td><p>Apply function over a List or Vector</p></td>
      </tr><tr><td>
          <p><code><a href="mean_module_expression.html">mean_module_expression()</a></code> </p>
        </td>
        <td><p>Computes mean expression of groups of genes</p></td>
      </tr><tr><td>
          <p><code><a href="mean_subnetwork_expression.html">mean_subnetwork_expression()</a></code> </p>
        </td>
        <td><p>Computes mean expression of groups of genes in a dynamic network</p></td>
      </tr><tr><td>
          <p><code><a href="no_legend.html">no_legend()</a></code> </p>
        </td>
        <td><p>Removes legend from plot.</p></td>
      </tr><tr><td>
          <p><code><a href="no_margin.html">no_margin()</a></code> </p>
        </td>
        <td><p>Removes margins from plot.</p></td>
      </tr><tr><td>
          <p><code><a href="no_x_text.html">no_x_text()</a></code> </p>
        </td>
        <td><p>Removes x axis text.</p></td>
      </tr><tr><td>
          <p><code><a href="no_y_text.html">no_y_text()</a></code> </p>
        </td>
        <td><p>Removes y axis text.</p></td>
      </tr><tr><td>
          <p><code><a href="order_genes.html">order_genes()</a></code> </p>
        </td>
        <td><p>Function that orders genes based on peak expression</p></td>
      </tr><tr><td>
          <p><code><a href="pre_pseudotime_matrix.html">pre_pseudotime_matrix()</a></code> </p>
        </td>
        <td><p>pre_pseudotime_matrix</p></td>
      </tr><tr><td>
          <p><code><a href="rough_hierarchy.html">rough_hierarchy()</a></code> </p>
        </td>
        <td><p>rough_hierarchy</p></td>
      </tr><tr><td>
          <p><code><a href="sample_and_epoch_reconstruct.html">sample_and_epoch_reconstruct()</a></code> </p>
        </td>
        <td><p>Function to bootstrap Epoch reconstruction</p></td>
      </tr><tr><td>
          <p><code><a href="score_targets.html">score_targets()</a></code> </p>
        </td>
        <td><p>Function to score targets of effectors</p></td>
      </tr><tr><td>
          <p><code><a href="sparse_cor.html">sparse_cor()</a></code> </p>
        </td>
        <td><p>Safe correlation function which returns a sparse matrix without missing values</p></td>
      </tr><tr><td>
          <p><code><a href="dMcast.html">dMcast()</a></code> </p>
        </td>
        <td><p>Copy of the dMcast function from the Matrix.utils package, since this is off CRAN and does not seem to be maintained anymore</p></td>
      </tr><tr><td>
          <p><code><a href="edge_rank.html">edge_rank()</a></code> </p>
        </td>
        <td><p>Function to compute edge differences</p></td>
      </tr><tr><td>
          <p><code><a href="fast_aggregate.html">fast_aggregate()</a></code> </p>
        </td>
        <td><p>Copy of the aggregate.Matrix function from the Matrix.utils package, since this is off CRAN and does not seem to be maintained anymore</p></td>
      </tr><tr><td>
          <p><code><a href="get_umap.html">get_umap()</a></code> </p>
        </td>
        <td><p>Compute UMAP embedding</p></td>
      </tr><tr><td>
          <p><code><a href="ig_NiceGraph.html">ig_NiceGraph()</a></code> </p>
        </td>
        <td><p>returns a pretty graph given a grnTab and expression data</p></td>
      </tr><tr><td>
          <p><code><a href="ig_convertLarge.html">ig_convertLarge()</a></code> </p>
        </td>
        <td><p>change igraph attributes so that it is suitable for plotting a small network</p></td>
      </tr><tr><td>
          <p><code><a href="ig_convertMedium.html">ig_convertMedium()</a></code> </p>
        </td>
        <td><p>change igraph attributes so that it is suitable for plotting a medium network</p></td>
      </tr><tr><td>
          <p><code><a href="ig_convertSmall.html">ig_convertSmall()</a></code> </p>
        </td>
        <td><p>change igraph attributes so that it is suitable for plotting a small network</p></td>
      </tr><tr><td>
          <p><code><a href="wn_ig_tabToIgraph.html">wn_ig_tabToIgraph()</a></code> </p>
        </td>
        <td><p>convert a table to an igraph</p></td>
      </tr><tr><td>
          <p><code><a href="ig_convertTFs.html">ig_convertTFs()</a></code> </p>
        </td>
        <td><p>change igraph attributes so that it is suitable for plotting a network of only regs</p></td>
      </tr><tr><td>
          <p><code><a href="ig_exemplars.html">ig_exemplars()</a></code> </p>
        </td>
        <td><p>make a graph of the regulators, top targets, selecting only top XX targets each</p></td>
      </tr><tr><td>
          <p><code><a href="ig_scaleV.html">ig_scaleV()</a></code> </p>
        </td>
        <td><p>return a vector of scaled sizes for a vector of verticies</p></td>
      </tr><tr><td>
          <p><code><a href="ig_tabToIgraph.html">ig_tabToIgraph()</a></code> </p>
        </td>
        <td><p>convert a table to an igraph</p></td>
      </tr><tr><td>
          <p><code><a href="log_message.html">log_message()</a></code> </p>
        </td>
        <td><p>Print diagnostic message</p></td>
      </tr><tr><td>
          <p><code><a href="r_square.html">r_square()</a></code> </p>
        </td>
        <td><p>\(R^2\) (coefficient of determination)</p></td>
      </tr><tr><td>
          <p><code><a href="rse.html">rse()</a></code> </p>
        </td>
        <td><p>Relative Squared Error</p></td>
      </tr><tr><td>
          <p><code><a href="sse.html">sse()</a></code> </p>
        </td>
        <td><p>Sum of Squared Errors</p></td>
      </tr><tr><td>
          <p><code><a href="sparse_covcor.html">sparse_covcor()</a></code> </p>
        </td>
        <td><p>Fast correlation and covariance calcualtion for sparse matrices</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="example-data">Example data <a href="#example-data" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="example_matrix.html">example_matrix</a></code> </p>
        </td>
        <td><p>Example matrix data</p></td>
      </tr><tr><td>
          <p><code><a href="example_ground_truth.html">example_ground_truth</a></code> </p>
        </td>
        <td><p>Example ground truth data</p></td>
      </tr><tr><td>
          <p><code><a href="example_meta_data.html">example_meta_data</a></code> </p>
        </td>
        <td><p>Eexample_meta_data</p></td>
      </tr><tr><td>
          <p><code><a href="promoter_regions_hg19.html">promoter_regions_hg19</a></code> </p>
        </td>
        <td><p>promoter regions data</p></td>
      </tr><tr><td>
          <p><code><a href="promoter_regions_hg38.html">promoter_regions_hg38</a></code> </p>
        </td>
        <td><p>promoter regions data</p></td>
      </tr><tr><td>
          <p><code><a href="EnsDb.Hsapiens.v93.annot.UCSC.hg38.html">EnsDb.Hsapiens.v93.annot.UCSC.hg38</a></code> </p>
        </td>
        <td><p>EnsDb.Hsapiens.v93.annot.UCSC.hg38</p></td>
      </tr><tr><td>
          <p><code><a href="SCREEN.ccRE.UCSC.hg38.html">SCREEN.ccRE.UCSC.hg38</a></code> </p>
        </td>
        <td><p>SCREEN.ccRE.UCSC.hg38</p></td>
      </tr><tr><td>
          <p><code><a href="phastConsElements20Mammals.UCSC.hg38.html">phastConsElements20Mammals.UCSC.hg38</a></code> </p>
        </td>
        <td><p>phastConsElements20Mammals.UCSC.hg38</p></td>
      </tr><tr><td>
          <p><code><a href="motif2tf.html">motif2tf</a></code> </p>
        </td>
        <td><p>motif2tf</p></td>
      </tr><tr><td>
          <p><code><a href="motifs.html">motifs</a></code> </p>
        </td>
        <td><p>motifs</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Meng Xu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

